Fix Content.Source InputStreamContentSource fillBufferFromInputStream

kohlschuetter opened the pull request:
Previously, that method would fail to correctly handle downloads of large InputStreams returning at least 4 GiB when the specified length was -1.

Address several boundary checks and restructure the code a little bit.

Also document the use of "-1" indicating "full length" in the javadoc.

Fixes #13196

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -293,9 +293,15 @@ protected int fillBufferFromInputStream(InputStream inputStream, byte[] buffer)
                         return -1;
-                    int toReadInt = (int)Math.min(Integer.MAX_VALUE, toRead);
-                    int len = toReadInt > -1 ? Math.min(toReadInt, buffer.length) : buffer.length;
+                    if (toRead < 0 || toRead > Integer.MAX_VALUE)
+                        return inputStream.read(buffer, 0, buffer.length);
```
shouldn't toRead be reduced by how many bytes are read here?

gregw reviewed (changes requested):
Note that this has been reworked in 12.1 and thus this fix should not be merged forward

lorban reviewed (changes requested):
This fix is missing a test.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -293,9 +293,15 @@ protected int fillBufferFromInputStream(InputStream inputStream, byte[] buffer)
+                    {
+                      return -1;
+                    }
                     toRead -= read;
```
```suggestion
                    if (read > 0)
                        toRead -= read;
```

sbordet commented:
@gregw note that also 12.1.x had bad handling of the correspondent code, so I opened #13206.

kohlschuetter committed:
Fix Content.Source InputStreamContentSource fillBufferFromInputStream

Previously, that method would fail to correctly handle downloads of
large InputStreams returning at least 4 GiB when the specified length
was -1.

Add boundary checks.

Also document the use of "-1" indicating "full length" in the javadoc.

Fixes #13196

Signed-off-by: Christian Kohlschütter <christian@kohlschutter.com>
